### PR TITLE
New git-prune-branches functionality

### PR DESCRIPTION
**Pull request by kevgo**
@charlierudolph @allewun 

Implements #702

### CODE REVIEW

File: src/helpers/git_helpers/branch_helpers.sh [hunk 1]
```diff
@@ -14,6 +14,13 @@ function branch_needs_push {
 
 
 # Checkout a branch
+function checkout_branch {
+  local branch_name=$1
+  run_command "git checkout $branch_name"
+}
```
Comment (charlierudolph) [inline src/helpers/git_helpers/branch_helpers.sh]: This is already implemented in `checkout_helpers.sh`. The function is just called `checkout`

File: features/git-prune-branches/normal_branch.feature [hunk 2]
```diff
@@ -0,0 +1,48 @@
+Feature: git prune-branches: delete branches that were shipped or removed on another machine
+
+  As a developer checking out branches that are also developed on another machine
+  I want to remove all branches that have been shipped or deleted on another machine
+  So that I keep my local repository free from obsolete branches and remain efficient.
+
+  Rules:
+  - branches that have the remote branch name "[origin/<branch name>: gone]" when running "git branch -vv" are obsolete
```
Comment (charlierudolph) [inline features/git-prune-branches/normal_branch.feature]: how about 
```cucumber
Rules:
- branches with a deleted tracking branch are removed
  - "git branch -vv" shows these branches with the remote branch name as "[origin/<branch name>: gone]"
```

File: src/git-prune-branches [hunk 3]
```diff
@@ -7,59 +7,21 @@ function preconditions {
...
-  for remote_branch_name in $(remote_only_merged_branches); do
-    if [ "$(should_delete_branch "$remote_branch_name")" = true ]; then
-      echo "delete_remote_only_branch $remote_branch_name"
-      echo_if_true "delete_parent_entry $remote_branch_name" "$(knows_parent_branch "$remote_branch_name")"
-    fi
-  done
+    # remove the branch from the branch hierarchy configuration
+    parent_branch_name=$(parent_branch "$branch_to_delete")
+    echo_update_child_branches "$branch_to_delete" "$parent_branch_name"
+    echo_if_true "delete_parent_entry $branch_to_delete" "$(knows_parent_branch "$branch_to_delete")"
```
Comment (charlierudolph) [inline src/git-prune-branches]: Seems odd that you update the child branches no matter what. What do you think about
```sh
if [ "$(knows_parent_branch "$branch_to_delete")" = true ]; then  
  local parent_branch_name=$(parent_branch "$branch_to_delete")
  echo_update_child_branches "$branch_to_delete" "$parent_branch_name"
  echo "delete_parent_entry $branch_to_delete"
fi
```

File: src/helpers/git_helpers/branch_helpers.sh [hunk 4]
```diff
@@ -175,6 +182,15 @@ function local_branches_without_main {
 }
 
 
+function local_branches_with_gone_remote {
+  git branch -vv              |  # get all branches
+  tr '*' ' '                  |  # remove the active branch marker
+  grep  "\[origin/.*: gone\]" |  # leave only branches whose remove is marked as gone
+  sed "s/^ *//"               |  # remove trailing whitespace
```
Comment (charlierudolph) [inline src/helpers/git_helpers/branch_helpers.sh]: This appears to actually be removing leading whitespace

File: src/helpers/git_helpers/branch_helpers.sh [hunk 5]
```diff
@@ -175,6 +182,15 @@ function local_branches_without_main {
 }
 
 
+function local_branches_with_gone_remote {
```
Comment (charlierudolph) [inline src/helpers/git_helpers/branch_helpers.sh]: What do you think about `local_branches_with_deleted_tracking_branch`?

File: src/helpers/git_helpers/branch_helpers.sh [hunk 6]
```diff
@@ -175,6 +182,15 @@ function local_branches_without_main {
...
+  git branch -vv              |  # get all branches
+  tr '*' ' '                  |  # remove the active branch marker
+  grep  "\[origin/.*: gone\]" |  # leave only branches whose remove is marked as gone
```
Comment (charlierudolph) [inline src/helpers/git_helpers/branch_helpers.sh]: remove -> remote
Would also prefer to use the delete tracking branch terminology though

Comment (charlierudolph) [inline src/helpers/git_helpers/branch_helpers.sh]: maybe "remote tracking branch" is the best terminology

**Comment by charlierudolph** [comment]
Awesome. Thanks for implementing this @kevgo

File: features/git-prune-branches/normal_branch.feature (hunk 2 above)
Comment (kevgo) [inline features/git-prune-branches/normal_branch.feature]: I like it!

File: src/git-prune-branches (hunk 3 above)
Comment (kevgo) [inline src/git-prune-branches]: Good catch. This is much better.

File: src/helpers/git_helpers/branch_helpers.sh (hunk 1 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_helpers.sh]: Ah, I was wondering about this. This is inconsistent with our naming strategy, and this file also contains a method `checkout_branch_silently`. I'm removing my changes here, and will also move the checkout_silently method over.

File: src/helpers/git_helpers/branch_helpers.sh (hunk 5 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_helpers.sh]: I like it. Adopted

File: src/helpers/git_helpers/branch_helpers.sh (hunk 6 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_helpers.sh]: I think "remote" and "tracking" are synonyms here. Afaik a tracking branch is always remote. I'd prefer to stick with the shorter version "tracking branch", until we also have non-remote tracking branches.

File: src/helpers/git_helpers/branch_helpers.sh (hunk 4 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_helpers.sh]: I always mix those two up  :)

**Comment by kevgo** [comment]
Thanks for the review! All comments addressed.

File: documentation/commands/git-prune-branches.md [hunk 7]
```diff
@@ -12,4 +12,6 @@ git prune-branches
 
 #### DESCRIPTION
 
-Deletes merged branches in the local and remote repositories.
+Deletes branches whose remote branch no longer exists,
```
Comment (charlierudolph) [inline documentation/commands/git-prune-branches.md]: This appears to be the last place where "tracking" might be "better" then remote for consistency.

File: documentation/commands/git-prune-branches.md [hunk 8]
```diff
@@ -12,4 +12,6 @@ git prune-branches
...
+i.e. branches that were shipped or killed on another machine,
+from the local repository.
```
Comment (charlierudolph) [inline documentation/commands/git-prune-branches.md]: Thoughts on rearranging this slightly so we don't have the clarification in the middle of the sentence.
```
Deletes branches whose remote branch no longer exists from the local repository.
(i.e. branches that were shipped or killed on another machine)
```

**Comment by charlierudolph** [comment]
Looks good to me after the final comments are addressed

File: documentation/commands/git-prune-branches.md (hunk 7 above)
Comment (kevgo) [inline documentation/commands/git-prune-branches.md]: Good idea. Changed.

File: documentation/commands/git-prune-branches.md (hunk 8 above)
Comment (kevgo) [inline documentation/commands/git-prune-branches.md]: What do you think about this:

```
Deletes branches whose tracking branch no longer exists from the local repository.
This usually means the branch was shipped or killed on another machine.
```

**Pull request closed without merging**